File: src/csharp/Foundgine.Core/Serialization/JsonReadIntentAdapter.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using Foundgine.Core.Semantic.Intent;
using Foundgine.Core.Semantic.Query;

namespace Foundgine.Core.Serialization;

/// <summary>
/// Translates a deliberately small JSON representation into Foundgine's
/// provider-neutral <see cref="ReadIntent"/>. It performs no semantic
/// resolution, authorization, planning, or provider work.
/// </summary>
public sealed class JsonReadIntentAdapter
{
    private readonly JsonReadIntentAdapterOptions _limits;

    public JsonReadIntentAdapter(JsonReadIntentAdapterOptions? limits = null)
    {
        _limits = limits ?? new JsonReadIntentAdapterOptions();
        ValidateLimits(_limits);
    }

    private static readonly JsonSerializerOptions Options = new()
    {
        PropertyNameCaseInsensitive = true,
        ReadCommentHandling = JsonCommentHandling.Skip,
        AllowTrailingCommas = true,
        UnmappedMemberHandling = JsonUnmappedMemberHandling.Disallow,
        Converters = { new JsonStringEnumConverter() }
    };

    private static readonly JsonSerializerOptions PermissiveOptions = new(Options)
    {
        UnmappedMemberHandling = JsonUnmappedMemberHandling.Skip
    };

    public ReadIntent Parse(string json)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(json);

        try
        {
            var options = _limits.RejectUnknownProperties ? Options : PermissiveOptions;
            var dto = JsonSerializer.Deserialize<ReadIntentDto>(json, options)
                      ?? throw Invalid("JSON document is empty.");
            return ToIntent(dto);
        }
        catch (JsonException ex)
        {
            throw Invalid($"Invalid JSON read intent: {ex.Message}", ex);
        }
    }

    private ReadIntent ToIntent(ReadIntentDto dto)
    {
        if (string.IsNullOrWhiteSpace(dto.RootEntity))
            throw Invalid("'rootEntity' is required.");

        if (dto.Selections is null || dto.Selections.Count == 0)
            throw Invalid("'selections' must contain at least one selection.");

        if (CountSelections(dto.Selections) > _limits.MaxSelections)
            throw Invalid($"Selection count exceeds the configured maximum of {_limits.MaxSelections}.");

        if (dto.Limit is < 0 || dto.Offset is < 0)
            throw Invalid("'limit' and 'offset' cannot be negative.");

        return new ReadIntent(
            dto.RootEntity,
            dto.Selections.Select(ToSelection).ToArray(),
            dto.Filter is null ? null : ToFilter(dto.Filter),
            dto.Order?.Select(ToOrder).ToArray(),
            dto.Limit,
            dto.Offset,
            dto.After);
    }

    private ReadSelection ToSelection(SelectionDto dto, int depth = 1)
    {
        if (depth > _limits.MaxSelectionDepth)
            throw Invalid($"Selection depth exceeds the configured maximum of {_limits.MaxSelectionDepth}.");

        var hasField = !string.IsNullOrWhiteSpace(dto.Field);
        var hasRelationship = !string.IsNullOrWhiteSpace(dto.Relationship);
        if (hasField == hasRelationship)
            throw Invalid("Each selection must specify exactly one of 'field' or 'relationship'.");

        return new ReadSelection(
            hasField ? dto.Field : null,
            hasRelationship ? dto.Relationship : null,
            dto.Children?.Select(child => ToSelection(child, depth + 1)).ToArray());
    }

    private ReadFilter ToFilter(FilterDto dto)
    {
        var nodes = 0;
        return ToFilter(dto, 1, ref nodes);
    }

    private ReadFilter ToFilter(FilterDto dto, int depth, ref int nodes)
    {
        if (depth > _limits.MaxFilterDepth)
            throw Invalid($"Filter depth exceeds the configured maximum of {_limits.MaxFilterDepth}.");
        if (++nodes > _limits.MaxFilterNodes)
            throw Invalid($"Filter node count exceeds the configured maximum of {_limits.MaxFilterNodes}.");
        if (dto.Kind is null)
            throw Invalid("Every filter requires a 'kind'.");

        return dto.Kind.Trim().ToLowerInvariant() switch
        {
            "field" => new ReadFieldFilter(
                Required(dto.Field, "filter.field"),
                dto.Operator ?? throw Invalid("Field filters require 'operator'."),
                ToValue(dto.Value, 0)),

            "relationship" => new ReadRelationshipFilter(
                Required(dto.Relationship, "filter.relationship"),
                dto.Quantifier ?? throw Invalid("Relationship filters require 'quantifier'."),
                dto.Predicate is null
                    ? throw Invalid("Relationship filters require 'predicate'.")
                    : ToFilter(dto.Predicate, depth + 1, ref nodes)),

            "and" => new ReadAndFilter(
                dto.Expressions is { Count: > 0 }
                    ? ToFilterList(dto.Expressions, depth + 1, ref nodes)
                    : throw Invalid("AND filters require at least one expression.")),

            "or" => new ReadOrFilter(
                dto.Expressions is { Count: > 0 }
                    ? ToFilterList(dto.Expressions, depth + 1, ref nodes)
                    : throw Invalid("OR filters require at least one expression.")),

            _ => throw Invalid($"Unsupported filter kind '{dto.Kind}'.")
        };
    }

    // A ref parameter cannot be captured inside a lambda (CS1628), so the
    // recursive per-expression compilation is done via an explicit loop
    // instead of dto.Expressions.Select(...).
    private ReadFilter[] ToFilterList(IReadOnlyList<FilterDto> expressions, int depth, ref int nodes)
    {
        var result = new ReadFilter[expressions.Count];
        for (var i = 0; i < expressions.Count; i++)
            result[i] = ToFilter(expressions[i], depth, ref nodes);
        return result;
    }

    private object? ToValue(JsonElement? value, int depth) => value is null ? null : Normalize(value.Value, depth);

    private object? Normalize(JsonElement value, int depth)
    {
        if (depth > _limits.MaxJsonValueDepth)
            throw Invalid($"JSON value depth exceeds the configured maximum of {_limits.MaxJsonValueDepth}.");

        return value.ValueKind switch
        {
            JsonValueKind.Null => null,
            JsonValueKind.True => true,
            JsonValueKind.False => false,
            JsonValueKind.String => value.GetString(),
            JsonValueKind.Number when value.TryGetInt64(out var integer) => integer,
            JsonValueKind.Number when value.TryGetDecimal(out var decimalValue) => decimalValue,
            JsonValueKind.Array => value.EnumerateArray().Select(item => Normalize(item, depth + 1)).ToArray(),
            JsonValueKind.Object => value.EnumerateObject()
                .ToDictionary(x => x.Name, x => Normalize(x.Value, depth + 1)),
            _ => throw Invalid($"Unsupported JSON value kind '{value.ValueKind}'.")
        };
    }

    private static ReadOrder ToOrder(OrderDto dto) =>
        new(
            Required(dto.Field, "order.field"),
            dto.Direction ?? throw Invalid("Order entries require 'direction'."),
            dto.RelationshipPath,
            dto.Aggregate ?? SemanticOrderAggregate.None);

    private int CountSelections(IEnumerable<SelectionDto> selections) =>
        CountSelections(selections, 1);

    private int CountSelections(IEnumerable<SelectionDto> selections, int depth)
    {
        if (depth > _limits.MaxSelectionDepth)
            throw Invalid($"Selection depth exceeds the configured maximum of {_limits.MaxSelectionDepth}.");

        var count = 0;
        foreach (var selection in selections)
        {
            if (++count > _limits.MaxSelections)
                return count;

            if (selection.Children is not null)
            {
                count += CountSelections(selection.Children, depth + 1);
                if (count > _limits.MaxSelections)
                    return count;
            }
        }

        return count;
    }

    private static void ValidateLimits(JsonReadIntentAdapterOptions limits)
    {
        if (limits.MaxSelectionDepth < 1 || limits.MaxSelections < 1 ||
            limits.MaxFilterDepth < 1 || limits.MaxFilterNodes < 1 ||
            limits.MaxJsonValueDepth < 1)
            throw new ArgumentOutOfRangeException(nameof(limits), "All parser limits must be positive.");
    }

    private static string Required(string? value, string name) =>
        !string.IsNullOrWhiteSpace(value) ? value : throw Invalid($"'{name}' is required.");

    private static InvalidOperationException Invalid(string message, Exception? inner = null) =>
        new(message, inner);

    private sealed class ReadIntentDto
    {
        public string? RootEntity { get; set; }
        public List<SelectionDto>? Selections { get; set; }
        public FilterDto? Filter { get; set; }
        public List<OrderDto>? Order { get; set; }
        public int? Limit { get; set; }
        public int? Offset { get; set; }
        public string? After { get; set; }
    }

    private sealed class SelectionDto
    {
        public string? Field { get; set; }
        public string? Relationship { get; set; }
        public List<SelectionDto>? Children { get; set; }
    }

    private sealed class FilterDto
    {
        public string? Kind { get; set; }
        public string? Field { get; set; }
        public SemanticFilterOperator? Operator { get; set; }
        public JsonElement? Value { get; set; }
        public string? Relationship { get; set; }
        public SemanticRelationshipQuantifier? Quantifier { get; set; }
        public FilterDto? Predicate { get; set; }
        public List<FilterDto>? Expressions { get; set; }
    }

    private sealed class OrderDto
    {
        public string? Field { get; set; }
        public SemanticSortDirection? Direction { get; set; }
        public List<string>? RelationshipPath { get; set; }
        public SemanticOrderAggregate? Aggregate { get; set; }
    }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/serialization/JsonReadIntentAdapter.java
package com.foundgine.core.serialization;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.foundgine.core.semantic.intent.ReadAndFilter;
import com.foundgine.core.semantic.intent.ReadFieldFilter;
import com.foundgine.core.semantic.intent.ReadFilter;
import com.foundgine.core.semantic.intent.ReadIntent;
import com.foundgine.core.semantic.intent.ReadOrFilter;
import com.foundgine.core.semantic.intent.ReadOrder;
import com.foundgine.core.semantic.intent.ReadRelationshipFilter;
import com.foundgine.core.semantic.intent.ReadSelection;
import com.foundgine.core.semantic.query.SemanticFilterOperator;
import com.foundgine.core.semantic.query.SemanticOrderAggregate;
import com.foundgine.core.semantic.query.SemanticRelationshipQuantifier;
import com.foundgine.core.semantic.query.SemanticSortDirection;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Translates a deliberately small JSON representation into Foundgine's
 * provider-neutral {@link ReadIntent}. It performs no semantic resolution,
 * authorization, planning, or provider work.
 *
 * <p>This walks the parsed Jackson {@link JsonNode} tree directly and
 * converts it into the domain read-intent types — {@link #rejectUnknown}
 * performs an unknown-property check applied per object node against
 * that node's known property names, rejecting untrusted JSON that contains
 * unexpected fields.
 */
public final class JsonReadIntentAdapter {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static final Set<String> INTENT_PROPERTIES =
            Set.of("rootEntity", "selections", "filter", "order", "limit", "offset", "after");
    private static final Set<String> SELECTION_PROPERTIES = Set.of("field", "relationship", "children");
    private static final Set<String> FILTER_PROPERTIES =
            Set.of("kind", "field", "operator", "value", "relationship", "quantifier", "predicate", "expressions");
    private static final Set<String> ORDER_PROPERTIES = Set.of("field", "direction", "relationshipPath", "aggregate");

    private final JsonReadIntentAdapterOptions limits;

    public JsonReadIntentAdapter() {
        this(null);
    }

    public JsonReadIntentAdapter(JsonReadIntentAdapterOptions limits) {
        this.limits = limits == null ? new JsonReadIntentAdapterOptions() : limits;
        validateLimits(this.limits);
    }

    public ReadIntent parse(String json) {
        if (json == null || json.isBlank())
            throw new IllegalArgumentException("json must not be null or blank.");

        JsonNode root;
        try {
            root = MAPPER.readTree(json);
        } catch (JsonProcessingException e) {
            throw invalid("Invalid JSON read intent: " + e.getMessage(), e);
        }
        if (root == null || root.isNull() || root.isMissingNode())
            throw invalid("JSON document is empty.");

        return toIntent(asObject(root, "root document"));
    }

    private ReadIntent toIntent(ObjectNode dto) {
        rejectUnknown(dto, INTENT_PROPERTIES);

        String rootEntity = text(dto, "rootEntity");
        if (rootEntity == null || rootEntity.isBlank())
            throw invalid("'rootEntity' is required.");

        JsonNode selectionsNode = dto.get("selections");
        if (selectionsNode == null || !selectionsNode.isArray() || selectionsNode.isEmpty())
            throw invalid("'selections' must contain at least one selection.");

        if (countSelections((ArrayNode) selectionsNode, 1) > limits.maxSelections())
            throw invalid("Selection count exceeds the configured maximum of " + limits.maxSelections() + ".");

        List<ReadSelection> selections = new ArrayList<>();
        for (JsonNode child : selectionsNode)
            selections.add(toSelection(asObject(child, "selection"), 1));

        JsonNode filterNode = dto.get("filter");
        ReadFilter filter = filterNode == null || filterNode.isNull() ? null : toFilter(asObject(filterNode, "filter"));

        List<ReadOrder> order = null;
        JsonNode orderNode = dto.get("order");
        if (orderNode != null && orderNode.isArray()) {
            order = new ArrayList<>();
            for (JsonNode item : orderNode)
                order.add(toOrder(asObject(item, "order")));
        }

        Integer limit = intOrNull(dto, "limit");
        Integer offset = intOrNull(dto, "offset");
        if ((limit != null && limit < 0) || (offset != null && offset < 0))
            throw invalidArgument("'limit' and 'offset' cannot be negative.");

        String after = text(dto, "after");

        return new ReadIntent(rootEntity, selections, filter, order, limit, offset, after, null);
    }

    private ReadSelection toSelection(ObjectNode dto, int depth) {
        if (depth > limits.maxSelectionDepth())
            throw invalid("Selection depth exceeds the configured maximum of " + limits.maxSelectionDepth() + ".");
        rejectUnknown(dto, SELECTION_PROPERTIES);

        String field = text(dto, "field");
        String relationship = text(dto, "relationship");
        boolean hasField = field != null && !field.isBlank();
        boolean hasRelationship = relationship != null && !relationship.isBlank();
        if (hasField == hasRelationship)
            throw invalidArgument("Each selection must specify exactly one of 'field' or 'relationship'.");

        List<ReadSelection> children = List.of();
        JsonNode childrenNode = dto.get("children");
        if (childrenNode != null && childrenNode.isArray()) {
            List<ReadSelection> collected = new ArrayList<>();
            for (JsonNode child : childrenNode)
                collected.add(toSelection(asObject(child, "selection child"), depth + 1));
            children = collected;
        }

        return new ReadSelection(hasField ? field : null, hasRelationship ? relationship : null, children);
    }

    private ReadFilter toFilter(ObjectNode dto) {
        int[] nodes = {0};
        return toFilter(dto, 1, nodes);
    }

    private ReadFilter toFilter(ObjectNode dto, int depth, int[] nodes) {
        if (depth > limits.maxFilterDepth())
            throw invalid("Filter depth exceeds the configured maximum of " + limits.maxFilterDepth() + ".");
        if (++nodes[0] > limits.maxFilterNodes())
            throw invalid("Filter node count exceeds the configured maximum of " + limits.maxFilterNodes() + ".");
        rejectUnknown(dto, FILTER_PROPERTIES);

        String kind = text(dto, "kind");
        if (kind == null)
            throw invalid("Every filter requires a 'kind'.");

        return switch (kind.trim().toLowerCase(java.util.Locale.ROOT)) {
            case "field" -> new ReadFieldFilter(
                    required(text(dto, "field"), "filter.field"),
                    parseEnum(SemanticFilterOperator.class, text(dto, "operator"), "Field filters require 'operator'."),
                    toValue(dto.get("value"), 0));

            case "relationship" -> new ReadRelationshipFilter(
                    required(text(dto, "relationship"), "filter.relationship"),
                    parseEnum(SemanticRelationshipQuantifier.class, text(dto, "quantifier"),
                            "Relationship filters require 'quantifier'."),
                    dto.get("predicate") == null || dto.get("predicate").isNull()
                            ? throwInvalid("Relationship filters require 'predicate'.")
                            : toFilter(asObject(dto.get("predicate"), "filter.predicate"), depth + 1, nodes));

            case "and" -> new ReadAndFilter(requireExpressions(dto, depth, nodes, "AND"));

            case "or" -> new ReadOrFilter(requireExpressions(dto, depth, nodes, "OR"));

            default -> throw invalidArgument("Unsupported filter kind '" + kind + "'.");
        };
    }

    private List<ReadFilter> requireExpressions(ObjectNode dto, int depth, int[] nodes, String kindLabel) {
        JsonNode expressionsNode = dto.get("expressions");
        if (expressionsNode == null || !expressionsNode.isArray() || expressionsNode.isEmpty())
            throw invalidArgument(kindLabel + " filters require at least one expression.");

        List<ReadFilter> result = new ArrayList<>();
        for (JsonNode expression : expressionsNode)
            result.add(toFilter(asObject(expression, "filter expression"), depth + 1, nodes));
        return result;
    }

    private Object toValue(JsonNode value, int depth) {
        return value == null || value.isNull() ? null : normalize(value, depth);
    }

    private Object normalize(JsonNode value, int depth) {
        if (depth > limits.maxJsonValueDepth())
            throw invalidArgument("JSON value depth exceeds the configured maximum of " + limits.maxJsonValueDepth() + ".");

        switch (value.getNodeType()) {
            case NULL:
                return null;
            case BOOLEAN:
                return value.booleanValue();
            case STRING:
                return value.textValue();
            case NUMBER:
                if (value.isIntegralNumber() && value.canConvertToLong())
                    return value.longValue();
                return value.decimalValue();
            case ARRAY: {
                List<Object> items = new ArrayList<>();
                for (JsonNode item : value)
                    items.add(normalize(item, depth + 1));
                return items;
            }
            case OBJECT: {
                Map<String, Object> map = new LinkedHashMap<>();
                Iterator<Map.Entry<String, JsonNode>> fields = value.fields();
                while (fields.hasNext()) {
                    Map.Entry<String, JsonNode> entry = fields.next();
                    map.put(entry.getKey(), normalize(entry.getValue(), depth + 1));
                }
                return map;
            }
            default:
                throw invalid("Unsupported JSON value kind '" + value.getNodeType() + "'.");
        }
    }

    private ReadOrder toOrder(ObjectNode dto) {
        rejectUnknown(dto, ORDER_PROPERTIES);

        List<String> relationshipPath = null;
        JsonNode pathNode = dto.get("relationshipPath");
        if (pathNode != null && pathNode.isArray()) {
            relationshipPath = new ArrayList<>();
            for (JsonNode item : pathNode)
                relationshipPath.add(item.textValue());
        }

        String aggregateText = text(dto, "aggregate");
        SemanticOrderAggregate aggregate = aggregateText == null
                ? SemanticOrderAggregate.NONE
                : parseEnum(SemanticOrderAggregate.class, aggregateText, "Invalid 'aggregate' value.");

        return new ReadOrder(
                required(text(dto, "field"), "order.field"),
                parseEnum(SemanticSortDirection.class, text(dto, "direction"), "Order entries require 'direction'."),
                relationshipPath,
                aggregate);
    }

    private int countSelections(ArrayNode selections, int depth) {
        if (depth > limits.maxSelectionDepth())
            throw invalid("Selection depth exceeds the configured maximum of " + limits.maxSelectionDepth() + ".");

        int count = 0;
        for (JsonNode selectionNode : selections) {
            if (++count > limits.maxSelections())
                return count;

            JsonNode childrenNode = selectionNode.get("children");
            if (childrenNode != null && childrenNode.isArray()) {
                count += countSelections((ArrayNode) childrenNode, depth + 1);
                if (count > limits.maxSelections())
                    return count;
            }
        }

        return count;
    }

    private void rejectUnknown(ObjectNode dto, Set<String> known) {
        if (!limits.rejectUnknownProperties())
            return;
        Iterator<String> names = dto.fieldNames();
        while (names.hasNext()) {
            String name = names.next();
            if (!known.contains(name))
                throw invalid("Unrecognized property '" + name + "'.");
        }
    }

    private static void validateLimits(JsonReadIntentAdapterOptions limits) {
        if (limits.maxSelectionDepth() < 1 || limits.maxSelections() < 1
                || limits.maxFilterDepth() < 1 || limits.maxFilterNodes() < 1
                || limits.maxJsonValueDepth() < 1)
            throw new IllegalArgumentException("All parser limits must be positive.");
    }

    private static ObjectNode asObject(JsonNode node, String context) {
        if (node == null || !node.isObject())
            throw invalid("'" + context + "' must be a JSON object.");
        return (ObjectNode) node;
    }

    private static String text(ObjectNode dto, String field) {
        JsonNode node = dto.get(field);
        return node == null || node.isNull() ? null : node.asText();
    }

    private static Integer intOrNull(ObjectNode dto, String field) {
        JsonNode node = dto.get(field);
        return node == null || node.isNull() ? null : node.asInt();
    }

    private static String required(String value, String name) {
        if (value == null || value.isBlank())
            throw invalid("'" + name + "' is required.");
        return value;
    }

    private static <E extends Enum<E>> E parseEnum(Class<E> type, String value, String missingMessage) {
        if (value == null || value.isBlank())
            throw invalid(missingMessage);
        try {
            return Enum.valueOf(type, value.trim().toUpperCase(java.util.Locale.ROOT));
        } catch (IllegalArgumentException e) {
            throw invalidArgument("Unsupported value '" + value + "' for " + type.getSimpleName() + ".");
        }
    }

    private static IllegalStateException invalid(String message) {
        return new IllegalStateException(message);
    }

    /**
     * Thrown for malformed-shape violations of the read intent itself (the
     * caller's JSON does not describe a structurally valid intent) — mirrors
     * this port's convention of using {@link IllegalArgumentException} for
     * argument-shape validation, matching the exception domain records such
     * as {@link com.foundgine.core.semantic.intent.ReadSelection} already
     * throw for the same class of violation. This is distinct from
     * {@link #invalid(String)}, which is reserved for untrusted-transport
     * security-boundary rejections (e.g. unrecognized/security-authority
     * properties) and resource-limit enforcement, both ported from this
     * codebase's {@code InvalidOperationException} → {@code IllegalStateException}
     * mapping.
     */
    private static IllegalArgumentException invalidArgument(String message) {
        return new IllegalArgumentException(message);
    }

    private static IllegalStateException invalid(String message, Throwable inner) {
        return new IllegalStateException(message, inner);
    }

    private static ReadFilter throwInvalid(String message) {
        throw invalid(message);
    }
}